### PR TITLE
Update to AGP 8.0

### LINKED ISSUE #660: [Bug]: Update to AGP 8.0
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is there a StackOverflow question about this issue?

- [X] I have searched StackOverflow

### What happened?

AGP 8.0 is stable with Android Studio Flamingo, so we should update our dependencies to it.

### Relevant logcat output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by alexvanyo**
Fixes #660 and #529

### CODE REVIEW

File: build-logic/convention/build.gradle.kts [hunk 1]
```diff
@@ -25,6 +27,12 @@ java {
     targetCompatibility = JavaVersion.VERSION_11
```
Comment (mlykotom) [inline build-logic/convention/build.gradle.kts:27]: you can raise the JavaVersion to 17. Here and in other places as well.

File: kokoro/build.sh [hunk 2]
```diff
@@ -36,7 +36,7 @@ echo y | ${ANDROID_HOME}/tools/bin/sdkmanager --licenses
 cd $KOKORO_ARTIFACTS_DIR/git/nowinandroid
 
 # The build needs Java 11, set it as the default Java version.
```
Comment (mlykotom) [inline kokoro/build.sh:38]:
```suggestion
# The build needs Java 17, set it as the default Java version.
```

**Review by mlykotom — APPROVED** [review]
You can raise the Java version to 17, otherwise LGTM

File: kokoro/build.sh (hunk 2 above)
Comment (alexvanyo) [reply kokoro/build.sh]: Good catch, updated.

File: build-logic/convention/build.gradle.kts (hunk 1 above)
Comment (alexvanyo) [reply build-logic/convention/build.gradle.kts]: Sounds good, raised to 17 everywhere.

File: build-logic/convention/src/main/kotlin/com/google/samples/apps/nowinandroid/KotlinAndroid.kt [hunk 3]
```diff
@@ -40,27 +40,28 @@ internal fun Project.configureKotlinAndroid(
...
+            sourceCompatibility = JavaVersion.VERSION_17
+            targetCompatibility = JavaVersion.VERSION_17
             isCoreLibraryDesugaringEnabled = true
         }
+    }
 
+    // Use withType to workaround https://youtrack.jetbrains.com/issue/KT-55947
+    tasks.withType<KotlinCompile>().configureEach {
         kotlinOptions {
+            // Set JVM target to 11
```
Comment (LinX64) [inline build-logic/convention/src/main/kotlin/com/google/samples/apps/nowinandroid/KotlinAndroid.kt:52]: Shouldn't we change this to 17, and not 11?
You're using `VERSION_17` in the next line.

Comment (alexvanyo) [reply build-logic/convention/src/main/kotlin/com/google/samples/apps/nowinandroid/KotlinAndroid.kt]: Ah yep, good catch

File: app/proguard-rules.pro [hunk 4]
```diff
@@ -24,3 +24,13 @@
 
 # @Serializable and @Polymorphic are used at runtime for polymorphic serialization.
 -keepattributes RuntimeVisibleAnnotations,AnnotationDefault
+
+-dontwarn org.bouncycastle.jsse.BCSSLParameters
```
Comment (itsandreramon) [inline app/proguard-rules.pro:28]: just curious why these rules are needed

File: build-logic/convention/build.gradle.kts [hunk 5]
```diff
@@ -14,15 +14,23 @@
...
     `kotlin-dsl`
 }
 
 group = "com.google.samples.apps.nowinandroid.buildlogic"
 
 java {
-    sourceCompatibility = JavaVersion.VERSION_11
-    targetCompatibility = JavaVersion.VERSION_11
+    sourceCompatibility = JavaVersion.VERSION_17
+    targetCompatibility = JavaVersion.VERSION_17
```
Comment (carlonzo) [inline build-logic/convention/build.gradle.kts:27]: I understand that AGP now requires JDK 17 to run, but why would you bump the source/targetCompatibility, and jvmTarget to 17? is this required as well?

Comment (borisf) [reply build-logic/convention/build.gradle.kts:27]: Basically to have the latest versions, toolchains and formats. We talk about importance of updating early and often - https://www.droidcon.com/2022/08/03/5-ways-to-improve-your-android-build-productivity/

Comment (carlonzo) [reply build-logic/convention/build.gradle.kts:27]: Thanks about your answer. I dont doubt the fact is important to update early and often. Using the latest java toolchain makes sense, but I cant find resource to understand the benefits on bumping the source compatibility and kotlin jvm target here. I assume that the generated bytecode will also need to be desugared before to be dexed? 
are there any references to documentation about this part?

Comment (itsandreramon) [reply build-logic/convention/build.gradle.kts:27]: technically you still get compiler improvements depending on what your actually using. you can still use JDK19 and set targetCompatibility to 17

File: build-logic/convention/src/main/kotlin/com/google/samples/apps/nowinandroid/AndroidCompose.kt [hunk 6]
```diff
@@ -40,16 +44,18 @@ internal fun Project.configureAndroidCompose(
...
-        }
-
         dependencies {
             val bom = libs.findLibrary("androidx-compose-bom").get()
             add("implementation", platform(bom))
             add("androidTestImplementation", platform(bom))
         }
     }
+
+    tasks.withType<KotlinCompile>().configureEach {
```
Comment (saihemanth121) [inline build-logic/convention/src/main/kotlin/com/google/samples/apps/nowinandroid/AndroidCompose.kt:54]: Is setting freeCompilerArgs on tasks different from older way of setting it through kotlinOptions extension?

**Review by dturner — APPROVED** [review]

File: build-logic/convention/build.gradle.kts (hunk 5 above)
Comment (dturner) [reply build-logic/convention/build.gradle.kts:27]: In the interest of getting this change merged I will resolve this conversation. 

@lub0s thanks for your suggestion makes a lot of sense to me. I'll pass this on to the tools team for prioritisation.

File: build-logic/convention/src/main/kotlin/com/google/samples/apps/nowinandroid/AndroidCompose.kt (hunk 6 above)
Comment (dturner) [reply build-logic/convention/src/main/kotlin/com/google/samples/apps/nowinandroid/AndroidCompose.kt:54]: In the interest of getting this merged I will resolve this conversation, but have passed this on to the tools team. Hopefully they can respond.

Comment (SimonMarquis) [reply build-logic/convention/src/main/kotlin/com/google/samples/apps/nowinandroid/AndroidCompose.kt:54]: nit: you don't even need to call `.configureEach`.
The following code does the same thing 🙂 

```kotlin
    tasks.withType<KotlinCompile> {
        /* */
    }
```

**Pull request merged**